File: services/gateway/internal/api/feed_mix.go
package api

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
)

// feedMix is how much of the home feed each source of new material gets.
//
// Three numbers, because three is what a person can hold in their head while
// deciding: the channels I follow, more of what I already watch, and something
// I have not seen before. The ranker has nine reasons and they overlap; asking
// somebody to tune those would be asking them to tune a vocabulary rather than
// a feed.
//
// Held by the gateway, like the translator's settings, and sent down with the
// request — so recsys keeps no configuration of its own and there is no file
// for the two of them to disagree about.
//
// One setting for the household rather than one per viewer. There are two
// seeded accounts and no sign-up screen (CLAUDE.md §7), so per-person taste is
// a privacy nobody has asked for, bought with a table, a migration and an RPC.
// If that changes, only the line that loads this has to.
type feedMix struct {
	Subscribed int `json:"subscribedPercent"`
	Affinity   int `json:"affinityPercent"`
	Discovery  int `json:"discoveryPercent"`
}

// defaultFeedMix is what "Reset to default" returns to.
//
// Duplicated rather than imported: the gateway does not link the ranker, and a
// REST layer reaching into another service's use cases for a constant is the
// dependency this architecture exists to prevent.
//
// It was 25/60/15, which was the split the fixed quota it replaced produced. On
// a library where nearly everything comes from subscribed channels that turned
// out to be badly wrong — the affinity slot held twenty-five videos against
// three and a half thousand subscribed ones, so a 60% affinity share spent half
// of every page scraping that bucket's floor while videos scoring three times
// higher went unused. A default that leads somewhere bad is worse than no
// default, because it is where the reset button sends you.
var defaultFeedMix = feedMix{Subscribed: 60, Affinity: 20, Discovery: 20}

// The shares the sliders never divide, in whole percent.
//
// Sent to the browser rather than hard-coded there. The settings page used to
// carry its own copy and spent a release quoting 82% after the fresh-subscribed
// share took ten of it — every slider's "N of 24" readout was overstated by a
// seventh, on the one screen whose job is to say what the numbers mean.
//
// The fresh share is a setting now, so it is read from the ranking config
// rather than assumed.
const (
	shareContinueWatchingPercent  = 10
	shareRewatchPercent           = 8
	defaultFreshSubscribedPercent = 10
)

func (g *Gateway) fixedShares() map[string]int {
	fresh := defaultFreshSubscribedPercent
	if set := g.loadRanking().FreshSubscribedPercent; set != nil {
		fresh = *set
	}
	return map[string]int{
		"continueWatching": shareContinueWatchingPercent,
		"rewatch":          shareRewatchPercent,
		"freshSubscribed":  fresh,
	}
}

func (g *Gateway) feedMixPath() string {
	return filepath.Join(g.configDir, "feed-mix.json")
}

// loadFeedMix reads what was saved, falling back to the defaults.
//
// Any unreadable or nonsensical file gives the defaults rather than an error:
// this decides the order of a grid, and there is no version of "the feed will
// not load" that is better than "the feed looks like it always did".
func loadFeedMix(path string) feedMix {
	raw, err := os.ReadFile(path)
	if err != nil || len(raw) == 0 {
		return defaultFeedMix
	}
	var saved feedMix
	if err := json.Unmarshal(raw, &saved); err != nil {
		return defaultFeedMix
	}
	if !saved.valid() {
		return defaultFeedMix
	}
	return saved
}

// valid rejects only what cannot be acted on: negatives, and all-zero.
//
// All three at zero is the one combination with no sensible reading — it asks
// for a feed of nothing new at all, which is an empty page rather than a
// preference. Anything else is honoured by ratio, so a household that sets
// 3/2/1 gets the same feed as one that sets 50/33/17.
func (m feedMix) valid() bool {
	if m.Subscribed < 0 || m.Affinity < 0 || m.Discovery < 0 {
		return false
	}
	return m.Subscribed+m.Affinity+m.Discovery > 0
}

func saveFeedMix(path string, m feedMix) error {
	blob, err := json.MarshalIndent(m, "", "  ")
	if err != nil {
		return err
	}
	return withFileLock(path, func() error { return writeFileAtomic(path, blob) })
}

func (g *Gateway) handleGetFeedMix(w http.ResponseWriter, _ *http.Request) {
	mix := loadFeedMix(g.feedMixPath())
	writeJSON(w, http.StatusOK, map[string]any{
		"subscribedPercent": mix.Subscribed,
		"affinityPercent":   mix.Affinity,
		"discoveryPercent":  mix.Discovery,
		// Sent rather than duplicated in the browser, so "Reset to default" and
		// the note explaining the default cannot drift from what the server
		// actually falls back to.
		"defaults": map[string]any{
			"subscribedPercent": defaultFeedMix.Subscribed,
			"affinityPercent":   defaultFeedMix.Affinity,
			"discoveryPercent":  defaultFeedMix.Discovery,
		},
		// What is not up for division, so the page can work out what a slider is
		// a percentage *of* rather than guessing.
		"fixedShares": g.fixedShares(),
	})
}

func (g *Gateway) handleSaveFeedMix(w http.ResponseWriter, r *http.Request) {
	var submitted feedMix
	if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 4<<10)).Decode(&submitted); err != nil {
		http.Error(w, "bad body", http.StatusBadRequest)
		return
	}
	if !submitted.valid() {
		http.Error(w, "a feed of nothing is not a setting", http.StatusBadRequest)
		return
	}
	if err := saveFeedMix(g.feedMixPath(), submitted); err != nil {
		g.logger.Warn("feed mix save", "error", err)
		http.Error(w, "could not save", http.StatusServiceUnavailable)
		return
	}
	g.logger.Info("feed mix saved",
		"subscribed", submitted.Subscribed,
		"affinity", submitted.Affinity,
		"discovery", submitted.Discovery)
	writeJSON(w, http.StatusOK, map[string]any{
		"subscribedPercent": submitted.Subscribed,
		"affinityPercent":   submitted.Affinity,
		"discoveryPercent":  submitted.Discovery,
	})
}

File: web/src/features/settings/domain/feed-mix.ts
/**
 * How the home feed's new material is divided, in whole percent.
 *
 * Three numbers rather than the ranker's nine reasons. The reasons overlap —
 * a video can be unwatched *and* from a subscribed channel, and only one of
 * those decides which share it takes — so offering them as controls would be
 * offering a vocabulary rather than a feed. These three are mutually exclusive
 * by construction: subscribed, or not subscribed and matching this viewer's
 * taste, or neither.
 */
export interface FeedMix {
  subscribedPercent: number
  affinityPercent: number
  discoveryPercent: number
}

export type FeedMixKey = keyof FeedMix

export const FEED_MIX_KEYS: FeedMixKey[] = [
  'subscribedPercent',
  'affinityPercent',
  'discoveryPercent',
]

/**
 * The shares nobody divides here, and why they are not sliders.
 *
 * Finishing something you started, going back to something you finished, and
 * being told when a channel you follow posts are not tastes. Making them compete
 * with the sliders would mean a viewer who wants more discovery is asked to give
 * up the video they were halfway through, or to stop hearing about new uploads.
 *
 * Read from the server rather than written here. This file used to carry its own
 * copy and spent a release saying the sliders divided 82% of the page — true
 * until the fresh-subscribed share took ten of it, after which every "N of 24"
 * readout was overstated by a seventh. On the one screen whose job is to say
 * what the numbers mean.
 */
export interface FixedShares {
  continueWatching: number
  rewatch: number
  freshSubscribed: number
}

/** What the server falls back to, and what the page shows before it answers. */
export const FALLBACK_FIXED_SHARES: FixedShares = {
  continueWatching: 10,
  rewatch: 8,
  freshSubscribed: 10,
}

/** What the three sliders divide between them, given the shares that are spoken for. */
export function adjustablePercent(fixed: FixedShares): number {
  const left = 100 - fixed.continueWatching - fixed.rewatch - fixed.freshSubscribed
  return left > 0 ? left : 0
}

/** The window the ratios are applied over, matching the ranker's page size. */
export const FEED_WINDOW = 24

export function mixTotal(mix: FeedMix): number {
  return mix.subscribedPercent + mix.affinityPercent + mix.discoveryPercent
}

/**
 * Move one slider and let the other two absorb the difference.
 *
 * The three always add to a hundred, so something has to give. The remainder is
 * split in proportion to where the other two already were, which is the only
 * division that leaves their relationship to each other untouched — dragging
 * "subscribed" up should not quietly decide that you now prefer discovery to
 * affinity.
 *
 * When the other two are both at zero there is no ratio to preserve, so they
 * split what is left evenly. Anything else would have to invent a preference.
 */
export function setShare(mix: FeedMix, key: FeedMixKey, value: number): FeedMix {
  const target = clampPercent(value)
  const others = FEED_MIX_KEYS.filter((k) => k !== key)
  const remainder = 100 - target
  const othersTotal = others.reduce((sum, k) => sum + mix[k], 0)

  const next = { ...mix, [key]: target } as FeedMix
  if (othersTotal <= 0) {
    const half = Math.round(remainder / 2)
    next[others[0]] = half
    next[others[1]] = remainder - half
    return next
  }

  // The first is rounded and the second takes what is left, so the three always
  // add to exactly a hundred however the rounding falls.
  const first = Math.round((mix[others[0]] / othersTotal) * remainder)
  next[others[0]] = first
  next[others[1]] = remainder - first
  return next
}

function clampPercent(value: number): number {
  if (!Number.isFinite(value)) return 0
  return Math.min(100, Math.max(0, Math.round(value)))
}

/**
 * How many of the next twenty-four videos a share works out as.
 *
 * A percentage of a percentage of a page is two pieces of arithmetic nobody
 * should have to do in their head to find out that thirty per cent means seven
 * videos. Shown beside each slider for the same reason the storage page shows
 * gigabytes rather than a ratio.
 */
export function videosPerWindow(percent: number, fixed: FixedShares): number {
  return Math.round((percent / 100) * (adjustablePercent(fixed) / 100) * FEED_WINDOW)
}
